diff --git a/csrc/sm90/prefill/sparse/config.h b/csrc/sm90/prefill/sparse/config.h
index 7500566..bc0e972 100644
--- a/csrc/sm90/prefill/sparse/config.h
+++ b/csrc/sm90/prefill/sparse/config.h
@@ -15,7 +15,7 @@ namespace sm90::fwd {
 
 using namespace cute;
 
-template<int D_QK, bool HAVE_TOPK_LENGTH>
+template<int D_QK, bool HAVE_TOPK_LENGTH, int STATIC_TOPK_BLOCKS = -1>
 class KernelTemplate {
 public:
 
diff --git a/csrc/sm90/prefill/sparse/fwd.cu b/csrc/sm90/prefill/sparse/fwd.cu
index eb62b5d..a8cf889 100644
--- a/csrc/sm90/prefill/sparse/fwd.cu
+++ b/csrc/sm90/prefill/sparse/fwd.cu
@@ -19,6 +19,8 @@ void run_fwd_kernel(const SparseAttnFwdParams& params) {
     } else if (params.d_qk == 576) {
         if (have_topk_length) {
             sm90::fwd::run_fwd_phase1_kernel<576, true>(params);
+        } else if (params.topk == 512) {
+            sm90::fwd::run_fwd_phase1_kernel<576, false, 8>(params);
         } else {
             sm90::fwd::run_fwd_phase1_kernel<576, false>(params);
         }
diff --git a/csrc/sm90/prefill/sparse/instantiations/phase1_k576.cu b/csrc/sm90/prefill/sparse/instantiations/phase1_k576.cu
index f35db00..7a005bd 100644
--- a/csrc/sm90/prefill/sparse/instantiations/phase1_k576.cu
+++ b/csrc/sm90/prefill/sparse/instantiations/phase1_k576.cu
@@ -4,5 +4,6 @@
 namespace sm90::fwd {
 
 template void run_fwd_phase1_kernel<576, false>(const SparseAttnFwdParams& params);
+template void run_fwd_phase1_kernel<576, false, 8>(const SparseAttnFwdParams& params);
 
 }
diff --git a/csrc/sm90/prefill/sparse/phase1.cuh b/csrc/sm90/prefill/sparse/phase1.cuh
index bf2fff8..d195813 100644
--- a/csrc/sm90/prefill/sparse/phase1.cuh
+++ b/csrc/sm90/prefill/sparse/phase1.cuh
@@ -39,9 +39,9 @@ void tma_bulk_reduce_add(void const* src_ptr, void* dst_ptr, int32_t store_bytes
                      : "memory");
 }
 
-template<int D_QK, bool HAVE_TOPK_LENGTH>
+template<int D_QK, bool HAVE_TOPK_LENGTH, int STATIC_TOPK_BLOCKS>
 template<typename TMAParams>
-__device__ void KernelTemplate<D_QK, HAVE_TOPK_LENGTH>::devfunc(const SparseAttnFwdParams &params, const TMAParams &tma_params) {
+__device__ void KernelTemplate<D_QK, HAVE_TOPK_LENGTH, STATIC_TOPK_BLOCKS>::devfunc(const SparseAttnFwdParams &params, const TMAParams &tma_params) {
 #if (defined(__CUDA_ARCH__) && (__CUDA_ARCH__ == 900)) || (defined(__CLION_IDE__) || defined(__VSCODE_IDE__))
     const int q_h_idx = blockIdx.x % (params.h_q/B_H);
     const int s_q_idx = blockIdx.x / (params.h_q/B_H);
@@ -78,7 +78,9 @@ __device__ void KernelTemplate<D_QK, HAVE_TOPK_LENGTH>::devfunc(const SparseAttn
     __syncthreads();
     
     const int topk_length = HAVE_TOPK_LENGTH ? __ldg(params.topk_length + s_q_idx) : params.topk;
-    const int num_topk_blocks = HAVE_TOPK_LENGTH ? ku::ceil_div(topk_length, (int)B_TOPK) : (int)((unsigned int)params.topk/(unsigned int)B_TOPK);
+    const int num_topk_blocks = STATIC_TOPK_BLOCKS > 0 ? STATIC_TOPK_BLOCKS : (
+        HAVE_TOPK_LENGTH ? ku::ceil_div(topk_length, (int)B_TOPK) : (int)((unsigned int)params.topk/(unsigned int)B_TOPK)
+    );
 
     if (warpgroup_idx == 0 || warpgroup_idx == 1) {
         cutlass::arch::warpgroup_reg_alloc<216>();
@@ -131,7 +133,6 @@ __device__ void KernelTemplate<D_QK, HAVE_TOPK_LENGTH>::devfunc(const SparseAttn
         };
 
         auto online_softmax_and_rescale_o = [&](auto warpgroup_idx) {
-            plan.bar_is_kv_valid_ready.wait(cur_bar_wait_phase);
             constexpr bool IS_WG1 = std::is_same_v<decltype(warpgroup_idx), Warpgroup1>;
             const float scale = params.sm_scale_div_log2;
             float r_sM[2];
@@ -303,7 +304,7 @@ __device__ void KernelTemplate<D_QK, HAVE_TOPK_LENGTH>::devfunc(const SparseAttn
                 }
             };
             
-            CUTE_NO_UNROLL
+            #pragma unroll 4
             for (int block_idx = 0; block_idx < num_topk_blocks; block_idx += 2) {
                 Tensor sV0l = make_tensor(make_smem_ptr(plan.k[0].data()), SmemLayoutKTilesTransposed<4>{});
                 Tensor sV1l = make_tensor(make_smem_ptr(plan.k[1].data()), SmemLayoutKTilesTransposed<4>{});
@@ -394,7 +395,7 @@ __device__ void KernelTemplate<D_QK, HAVE_TOPK_LENGTH>::devfunc(const SparseAttn
                 warpgroup_commit_batch();
             };
             
-            CUTE_NO_UNROLL
+            #pragma unroll 4
             for (int block_idx = 0; block_idx < num_topk_blocks; block_idx += 2) {
                 Tensor sV0r = make_tensor(make_smem_ptr(plan.k[0].data()+64*256), SmemLayoutKTilesTransposed<4>{});
                 Tensor sV1r = make_tensor(make_smem_ptr(plan.k[1].data()+64*256), SmemLayoutKTilesTransposed<4>{});
@@ -518,7 +519,7 @@ __device__ void KernelTemplate<D_QK, HAVE_TOPK_LENGTH>::devfunc(const SparseAttn
 
         int cur_bar_wait_phase = 1;
 
-        CUTE_NO_UNROLL
+        #pragma unroll 4
         for (int block_idx = 0; block_idx < num_topk_blocks; block_idx += 2) {
             load_token_indices(block_idx);
 
@@ -571,12 +572,16 @@ sparse_attn_fwd_kernel(__grid_constant__ const SparseAttnFwdParams params, __gri
     Kernel::devfunc(params, tma_params);
 }
 
-template<int D_QK, bool HAVE_TOPK_LENGTH>
-void KernelTemplate<D_QK, HAVE_TOPK_LENGTH>::run(const SparseAttnFwdParams &params) {
+template<int D_QK, bool HAVE_TOPK_LENGTH, int STATIC_TOPK_BLOCKS>
+void KernelTemplate<D_QK, HAVE_TOPK_LENGTH, STATIC_TOPK_BLOCKS>::run(const SparseAttnFwdParams &params) {
     KU_ASSERT(params.h_kv == 1);
     KU_ASSERT(params.topk % (2*B_TOPK) == 0);   // To save some boundry checkings
     KU_ASSERT(params.topk > 0);
     KU_ASSERT(params.h_q % B_H == 0);
+    if constexpr (STATIC_TOPK_BLOCKS > 0) {
+        KU_ASSERT(!HAVE_TOPK_LENGTH);
+        KU_ASSERT(params.topk == STATIC_TOPK_BLOCKS * B_TOPK);
+    }
 
     auto shape_Q = make_shape(params.h_q, params.d_qk, params.s_q);
     auto tma_Q = cute::make_tma_copy(
@@ -620,7 +625,7 @@ void KernelTemplate<D_QK, HAVE_TOPK_LENGTH>::run(const SparseAttnFwdParams &para
         shape_Q, tma_Q,
         tensor_map_O
     };
-    auto kernel = &sparse_attn_fwd_kernel<KernelTemplate<D_QK, HAVE_TOPK_LENGTH>, decltype(tma_params)>;
+    auto kernel = &sparse_attn_fwd_kernel<KernelTemplate<D_QK, HAVE_TOPK_LENGTH, STATIC_TOPK_BLOCKS>, decltype(tma_params)>;
 
     constexpr size_t smem_size = sizeof(SharedMemoryPlan);
     KU_CUDA_CHECK(cudaFuncSetAttribute(kernel, cudaFuncAttributeMaxDynamicSharedMemorySize, smem_size));
@@ -638,9 +643,9 @@ void KernelTemplate<D_QK, HAVE_TOPK_LENGTH>::run(const SparseAttnFwdParams &para
     KU_CHECK_KERNEL_LAUNCH();
 }
 
-template<int D_QK, bool HAVE_TOPK_LENGTH>
+template<int D_QK, bool HAVE_TOPK_LENGTH, int STATIC_TOPK_BLOCKS>
 void run_fwd_phase1_kernel(const SparseAttnFwdParams& params) {
-    KernelTemplate<D_QK, HAVE_TOPK_LENGTH>::run(params);
+    KernelTemplate<D_QK, HAVE_TOPK_LENGTH, STATIC_TOPK_BLOCKS>::run(params);
 }
 
 }
diff --git a/csrc/sm90/prefill/sparse/phase1.h b/csrc/sm90/prefill/sparse/phase1.h
index c315b2b..3ba2bf2 100644
--- a/csrc/sm90/prefill/sparse/phase1.h
+++ b/csrc/sm90/prefill/sparse/phase1.h
@@ -4,7 +4,7 @@
 
 namespace sm90::fwd {
 
-template<int D_QK, bool HAVE_TOPK_LENGTH>
+template<int D_QK, bool HAVE_TOPK_LENGTH, int STATIC_TOPK_BLOCKS = -1>
 void run_fwd_phase1_kernel(const SparseAttnFwdParams& params);
 
 }
