diff --git a/csrc/sm90/prefill/sparse/config.h b/csrc/sm90/prefill/sparse/config.h
index 7500566..40f6d0d 100644
--- a/csrc/sm90/prefill/sparse/config.h
+++ b/csrc/sm90/prefill/sparse/config.h
@@ -15,7 +15,7 @@ namespace sm90::fwd {
 
 using namespace cute;
 
-template<int D_QK, bool HAVE_TOPK_LENGTH>
+template<int D_QK, bool HAVE_TOPK_LENGTH, bool ASSUME_VALID_INDICES = false>
 class KernelTemplate {
 public:
 
diff --git a/csrc/sm90/prefill/sparse/fwd.cu b/csrc/sm90/prefill/sparse/fwd.cu
index eb62b5d..b2ec9c4 100644
--- a/csrc/sm90/prefill/sparse/fwd.cu
+++ b/csrc/sm90/prefill/sparse/fwd.cu
@@ -19,6 +19,8 @@ void run_fwd_kernel(const SparseAttnFwdParams& params) {
     } else if (params.d_qk == 576) {
         if (have_topk_length) {
             sm90::fwd::run_fwd_phase1_kernel<576, true>(params);
+        } else if (params.topk == 512) {
+            sm90::fwd::run_fwd_phase1_kernel<576, false, true>(params);
         } else {
             sm90::fwd::run_fwd_phase1_kernel<576, false>(params);
         }
diff --git a/csrc/sm90/prefill/sparse/instantiations/phase1_k576.cu b/csrc/sm90/prefill/sparse/instantiations/phase1_k576.cu
index f35db00..04e6eda 100644
--- a/csrc/sm90/prefill/sparse/instantiations/phase1_k576.cu
+++ b/csrc/sm90/prefill/sparse/instantiations/phase1_k576.cu
@@ -4,5 +4,6 @@
 namespace sm90::fwd {
 
 template void run_fwd_phase1_kernel<576, false>(const SparseAttnFwdParams& params);
+template void run_fwd_phase1_kernel<576, false, true>(const SparseAttnFwdParams& params);
 
 }
diff --git a/csrc/sm90/prefill/sparse/phase1.cuh b/csrc/sm90/prefill/sparse/phase1.cuh
index bf2fff8..4a81240 100644
--- a/csrc/sm90/prefill/sparse/phase1.cuh
+++ b/csrc/sm90/prefill/sparse/phase1.cuh
@@ -39,9 +39,9 @@ void tma_bulk_reduce_add(void const* src_ptr, void* dst_ptr, int32_t store_bytes
                      : "memory");
 }
 
-template<int D_QK, bool HAVE_TOPK_LENGTH>
+template<int D_QK, bool HAVE_TOPK_LENGTH, bool ASSUME_VALID_INDICES>
 template<typename TMAParams>
-__device__ void KernelTemplate<D_QK, HAVE_TOPK_LENGTH>::devfunc(const SparseAttnFwdParams &params, const TMAParams &tma_params) {
+__device__ void KernelTemplate<D_QK, HAVE_TOPK_LENGTH, ASSUME_VALID_INDICES>::devfunc(const SparseAttnFwdParams &params, const TMAParams &tma_params) {
 #if (defined(__CUDA_ARCH__) && (__CUDA_ARCH__ == 900)) || (defined(__CLION_IDE__) || defined(__VSCODE_IDE__))
     const int q_h_idx = blockIdx.x % (params.h_q/B_H);
     const int s_q_idx = blockIdx.x / (params.h_q/B_H);
@@ -81,7 +81,7 @@ __device__ void KernelTemplate<D_QK, HAVE_TOPK_LENGTH>::devfunc(const SparseAttn
     const int num_topk_blocks = HAVE_TOPK_LENGTH ? ku::ceil_div(topk_length, (int)B_TOPK) : (int)((unsigned int)params.topk/(unsigned int)B_TOPK);
 
     if (warpgroup_idx == 0 || warpgroup_idx == 1) {
-        cutlass::arch::warpgroup_reg_alloc<216>();
+        cutlass::arch::warpgroup_reg_alloc<224>();
 
         if (warp_idx == 0 && elect_one_sync()) {
             // Load Q
@@ -117,21 +117,23 @@ __device__ void KernelTemplate<D_QK, HAVE_TOPK_LENGTH>::devfunc(const SparseAttn
         };
 
         auto mask_rP = [&](auto warpgroup_idx) {
-            constexpr bool IS_WG1 = std::is_same_v<decltype(warpgroup_idx), Warpgroup1>;
-            plan.bar_is_kv_valid_ready.wait(cur_bar_wait_phase);
-            CUTE_UNROLL
-            for (int row_idx = 0; row_idx < 2; ++row_idx) {
+            if constexpr (!ASSUME_VALID_INDICES) {
+                constexpr bool IS_WG1 = std::is_same_v<decltype(warpgroup_idx), Warpgroup1>;
+                plan.bar_is_kv_valid_ready.wait(cur_bar_wait_phase);
                 CUTE_UNROLL
-                for (int i = row_idx*2; i < size(rP); i += 4) {
-                    int col = 8*(i/4) + (idx_in_warpgroup%4)*2;
-                    if (!plan.is_kv_valid[IS_WG1][col]) rP(i) = -INFINITY;
-                    if (!plan.is_kv_valid[IS_WG1][col+1]) rP(i+1) = -INFINITY;
+                for (int row_idx = 0; row_idx < 2; ++row_idx) {
+                    CUTE_UNROLL
+                    for (int i = row_idx*2; i < size(rP); i += 4) {
+                        int col = 8*(i/4) + (idx_in_warpgroup%4)*2;
+                        if (!plan.is_kv_valid[IS_WG1][col]) rP(i) = -INFINITY;
+                        if (!plan.is_kv_valid[IS_WG1][col+1]) rP(i+1) = -INFINITY;
+                    }
                 }
             }
         };
 
         auto online_softmax_and_rescale_o = [&](auto warpgroup_idx) {
-            plan.bar_is_kv_valid_ready.wait(cur_bar_wait_phase);
+            if constexpr (!ASSUME_VALID_INDICES) plan.bar_is_kv_valid_ready.wait(cur_bar_wait_phase);
             constexpr bool IS_WG1 = std::is_same_v<decltype(warpgroup_idx), Warpgroup1>;
             const float scale = params.sm_scale_div_log2;
             float r_sM[2];
@@ -461,7 +463,7 @@ __device__ void KernelTemplate<D_QK, HAVE_TOPK_LENGTH>::devfunc(const SparseAttn
         }
     } else {
         // Producer warpgroup
-        cutlass::arch::warpgroup_reg_dealloc<72>();
+        cutlass::arch::warpgroup_reg_dealloc<56>();
 
         constexpr int GROUP_SIZE = 8, NUM_GROUPS = 128/GROUP_SIZE;
         constexpr int NUM_ROWS_PER_GROUP = B_TOPK / NUM_GROUPS;
@@ -482,11 +484,13 @@ __device__ void KernelTemplate<D_QK, HAVE_TOPK_LENGTH>::devfunc(const SparseAttn
                     int offs = (block_idx+buf_idx)*B_TOPK + local_row*NUM_GROUPS + group_idx;
                     int t = __ldg(gIndices + offs);
                     token_indices[buf_idx][local_row] = t*(int64_t)params.stride_kv_s_kv;   // We mult it with params.stride_kv_s_kv here since it's faster
-                    bool is_cur_token_valid = t >= 0 && t < params.s_kv;
-                    if constexpr (HAVE_TOPK_LENGTH) {
-                        is_cur_token_valid &= offs < topk_length;
+                    if constexpr (!ASSUME_VALID_INDICES) {
+                        bool is_cur_token_valid = t >= 0 && t < params.s_kv;
+                        if constexpr (HAVE_TOPK_LENGTH) {
+                            is_cur_token_valid &= offs < topk_length;
+                        }
+                        is_token_valid[buf_idx][local_row] = is_cur_token_valid;
                     }
-                    is_token_valid[buf_idx][local_row] = is_cur_token_valid;
                 }
             }
         };
@@ -502,10 +506,14 @@ __device__ void KernelTemplate<D_QK, HAVE_TOPK_LENGTH>::devfunc(const SparseAttn
                 int64_t token_index = token_indices[buf_idx][local_row];
                 CUTE_UNROLL
                 for (int tile_idx = tile_start; tile_idx < tile_end; ++tile_idx) {
+                    bool do_copy = true;
+                    if constexpr (!ASSUME_VALID_INDICES) {
+                        do_copy = is_token_valid[buf_idx][local_row];
+                    }
                     cp_async_cacheglobal_l2_prefetch_256B(
                         my_gKV_base + token_index + tile_idx*64,
                         my_sKV_base + (buf_idx*B_TOPK*D_K + tile_idx*(B_TOPK*64) + local_row*NUM_GROUPS*64),
-                        is_token_valid[buf_idx][local_row],
+                        do_copy,
                         cache_policy
                     );
                 }
@@ -527,16 +535,16 @@ __device__ void KernelTemplate<D_QK, HAVE_TOPK_LENGTH>::devfunc(const SparseAttn
             copy_tiles(block_idx+0, 0, 0, 4);
             commit_to_mbar(plan.bar_k0_ready[0]);
 
-            // V1R
-            plan.bar_k1_free[1].wait(cur_bar_wait_phase);
-            copy_tiles(block_idx+1, 1, 4, D_K/64);
-            commit_to_mbar(plan.bar_k1_ready[1]);
-            
             // V0R
             plan.bar_k0_free[1].wait(cur_bar_wait_phase);
             copy_tiles(block_idx+0, 0, 4, D_K/64);
             commit_to_mbar(plan.bar_k0_ready[1]);
 
+            // V1R
+            plan.bar_k1_free[1].wait(cur_bar_wait_phase);
+            copy_tiles(block_idx+1, 1, 4, D_K/64);
+            commit_to_mbar(plan.bar_k1_ready[1]);
+
             // V1L
             plan.bar_k1_free[0].wait(cur_bar_wait_phase);
             copy_tiles(block_idx+1, 1, 0, 4);
@@ -544,13 +552,15 @@ __device__ void KernelTemplate<D_QK, HAVE_TOPK_LENGTH>::devfunc(const SparseAttn
 
             // Valid mask
             // NOTE: V1R's finish implies maskings of the last round have finished
-            if (idx_in_group == 0) {
-                CUTE_UNROLL
-                for (int buf_idx = 0; buf_idx < 2; ++buf_idx)
+            if constexpr (!ASSUME_VALID_INDICES) {
+                if (idx_in_group == 0) {
                     CUTE_UNROLL
-                    for (int local_row = 0; local_row < NUM_ROWS_PER_GROUP; ++local_row)
-                        plan.is_kv_valid[buf_idx][local_row*NUM_GROUPS+group_idx] = is_token_valid[buf_idx][local_row];
-                plan.bar_is_kv_valid_ready.arrive();
+                    for (int buf_idx = 0; buf_idx < 2; ++buf_idx)
+                        CUTE_UNROLL
+                        for (int local_row = 0; local_row < NUM_ROWS_PER_GROUP; ++local_row)
+                            plan.is_kv_valid[buf_idx][local_row*NUM_GROUPS+group_idx] = is_token_valid[buf_idx][local_row];
+                    plan.bar_is_kv_valid_ready.arrive();
+                }
             }
 
             cur_bar_wait_phase ^= 1;
@@ -571,12 +581,17 @@ sparse_attn_fwd_kernel(__grid_constant__ const SparseAttnFwdParams params, __gri
     Kernel::devfunc(params, tma_params);
 }
 
-template<int D_QK, bool HAVE_TOPK_LENGTH>
-void KernelTemplate<D_QK, HAVE_TOPK_LENGTH>::run(const SparseAttnFwdParams &params) {
+template<int D_QK, bool HAVE_TOPK_LENGTH, bool ASSUME_VALID_INDICES>
+void KernelTemplate<D_QK, HAVE_TOPK_LENGTH, ASSUME_VALID_INDICES>::run(const SparseAttnFwdParams &params) {
     KU_ASSERT(params.h_kv == 1);
     KU_ASSERT(params.topk % (2*B_TOPK) == 0);   // To save some boundry checkings
     KU_ASSERT(params.topk > 0);
     KU_ASSERT(params.h_q % B_H == 0);
+    if constexpr (ASSUME_VALID_INDICES) {
+        KU_ASSERT(D_QK == 576);
+        KU_ASSERT(!HAVE_TOPK_LENGTH);
+        KU_ASSERT(params.topk == 512);
+    }
 
     auto shape_Q = make_shape(params.h_q, params.d_qk, params.s_q);
     auto tma_Q = cute::make_tma_copy(
@@ -620,7 +635,7 @@ void KernelTemplate<D_QK, HAVE_TOPK_LENGTH>::run(const SparseAttnFwdParams &para
         shape_Q, tma_Q,
         tensor_map_O
     };
-    auto kernel = &sparse_attn_fwd_kernel<KernelTemplate<D_QK, HAVE_TOPK_LENGTH>, decltype(tma_params)>;
+    auto kernel = &sparse_attn_fwd_kernel<KernelTemplate<D_QK, HAVE_TOPK_LENGTH, ASSUME_VALID_INDICES>, decltype(tma_params)>;
 
     constexpr size_t smem_size = sizeof(SharedMemoryPlan);
     KU_CUDA_CHECK(cudaFuncSetAttribute(kernel, cudaFuncAttributeMaxDynamicSharedMemorySize, smem_size));
@@ -638,9 +653,9 @@ void KernelTemplate<D_QK, HAVE_TOPK_LENGTH>::run(const SparseAttnFwdParams &para
     KU_CHECK_KERNEL_LAUNCH();
 }
 
-template<int D_QK, bool HAVE_TOPK_LENGTH>
+template<int D_QK, bool HAVE_TOPK_LENGTH, bool ASSUME_VALID_INDICES>
 void run_fwd_phase1_kernel(const SparseAttnFwdParams& params) {
-    KernelTemplate<D_QK, HAVE_TOPK_LENGTH>::run(params);
+    KernelTemplate<D_QK, HAVE_TOPK_LENGTH, ASSUME_VALID_INDICES>::run(params);
 }
 
 }
diff --git a/csrc/sm90/prefill/sparse/phase1.h b/csrc/sm90/prefill/sparse/phase1.h
index c315b2b..8084c58 100644
--- a/csrc/sm90/prefill/sparse/phase1.h
+++ b/csrc/sm90/prefill/sparse/phase1.h
@@ -4,7 +4,7 @@
 
 namespace sm90::fwd {
 
-template<int D_QK, bool HAVE_TOPK_LENGTH>
+template<int D_QK, bool HAVE_TOPK_LENGTH, bool ASSUME_VALID_INDICES = false>
 void run_fwd_phase1_kernel(const SparseAttnFwdParams& params);
 
 }
