diff --git a/csrc/sm90/prefill/sparse/phase1.cuh b/csrc/sm90/prefill/sparse/phase1.cuh
--- a/csrc/sm90/prefill/sparse/phase1.cuh
+++ b/csrc/sm90/prefill/sparse/phase1.cuh
@@ -84,7 +84,7 @@ __device__ void KernelTemplate<D_QK, HAVE_TOPK_LENGTH>::devfunc(const SparseAttn
     const int num_topk_blocks = HAVE_TOPK_LENGTH ? ku::ceil_div(topk_length, (int)B_TOPK) : (int)((unsigned int)params.topk/(unsigned int)B_TOPK);
 
     if (warpgroup_idx == 0 || warpgroup_idx == 1) {
-        cutlass::arch::warpgroup_reg_alloc<216>();
+        cutlass::arch::warpgroup_reg_alloc<224>();
 
         if (warp_idx == 0 && elect_one_sync()) {
             // Load Q
@@ -131,7 +131,6 @@ __device__ void KernelTemplate<D_QK, HAVE_TOPK_LENGTH>::devfunc(const SparseAttn
         };
 
         auto online_softmax_and_rescale_o = [&](auto warpgroup_idx) {
-            plan.bar_is_kv_valid_ready.wait(cur_bar_wait_phase);
             constexpr bool IS_WG1 = std::is_same_v<decltype(warpgroup_idx), Warpgroup1>;
             const float scale = params.sm_scale_div_log2;
             float r_sM[2];
@@ -463,7 +462,7 @@ __device__ void KernelTemplate<D_QK, HAVE_TOPK_LENGTH>::devfunc(const SparseAttn
         }
     } else {
         // Producer warpgroup
-        cutlass::arch::warpgroup_reg_dealloc<72>();
+        cutlass::arch::warpgroup_reg_dealloc<56>();
 
         constexpr int GROUP_SIZE = 8, NUM_GROUPS = 128/GROUP_SIZE;
         constexpr int NUM_ROWS_PER_GROUP = B_TOPK / NUM_GROUPS;
@@ -526,15 +525,15 @@ __device__ void KernelTemplate<D_QK, HAVE_TOPK_LENGTH>::devfunc(const SparseAttn
             copy_tiles(block_idx+0, 0, 0, 4);
             commit_to_mbar(plan.bar_k0_ready[0]);
 
-            // V1R
-            plan.bar_k1_free[1].wait(cur_bar_wait_phase);
-            copy_tiles(block_idx+1, 1, 4, D_K/64);
-            commit_to_mbar(plan.bar_k1_ready[1]);
-            
             // V0R
             plan.bar_k0_free[1].wait(cur_bar_wait_phase);
             copy_tiles(block_idx+0, 0, 4, D_K/64);
             commit_to_mbar(plan.bar_k0_ready[1]);
+
+            // V1R
+            plan.bar_k1_free[1].wait(cur_bar_wait_phase);
+            copy_tiles(block_idx+1, 1, 4, D_K/64);
+            commit_to_mbar(plan.bar_k1_ready[1]);
 
             // V1L
             plan.bar_k1_free[0].wait(cur_bar_wait_phase);
             copy_tiles(block_idx+1, 1, 0, 4);
             commit_to_mbar(plan.bar_k1_ready[0]);
